python/ray/data/_internal/util.py

```diff
@@ -421,6 +423,20 @@ def validate_compute(
             "For example, use ``compute=ray.data.ActorPoolStrategy(size=n)``."
         )
 
+    if fn_constructor_args is not None:
+        if compute is None or (
+            compute != "actors" and not isinstance(compute, ActorPoolStrategy)
+        ):
+            raise ValueError(
+                "fn_constructor_args can only be specified if using the actor "
+                f"pool compute strategy, but got: {compute}"
+            )
+        if not isinstance(fn, CallableClass):
+            raise ValueError(
+                "fn_constructor_args can only be specified if providing a "
+                f"CallableClass instance for fn, but got: {fn}"
+            )
```

Comment:
The type of the compute arg is too loose. This makes us need duplicated null checks.

It'd be better to split this into 2 steps. 
1) a `parse_compute` util function that converts `Optional[Union[str, "ComputeStrategy"]]` to just `ComputeStrategy`. 
2) Then check the fn and args. 

And by doing this, `compute` can only be None or str in the Dataset API level. under the hood, it's always the concrete type.

Reply:
Make sense, we already have a [`get_compute()`](https://github.com/ray-project/ray/blob/master/python/ray/data/_internal/compute.py#L461) method.

Let me make a separate PR to do the refactoring, it would involve change `AbstractUDFMap.compute` and `_plan_udf_map_op`.